### PR TITLE
P3-233 Change Republish Copies

### PR DESCRIPTION
**Pull request by karlijnbok**
## Context

If you are editing a post that has been intended for republishing then all buttons on the post edit screen that mention publishing should instead mention republishing.

The following copies have been changed:

_Block editor_
* `Publish` to `Republish` in the top right corner.
Screenshot 2020-12-14 at 11 26 02

* `Publish` to `Republish` in the pre-publish screen.
Screenshot 2020-12-14 at 11 26 39

* Text in the pre-publish screen. 
Screenshot 2020-12-14 at 11 26 55

* `Schedule` to `Schedule republish` in the top right corner.
Screenshot 2020-12-14 at 11 27 17

* `Schedule` to `Schedule republish` in the pre-publish screen.
Screenshot 2020-12-14 at 11 27 57

* Text in schedule pre-publish screen. 
Screenshot 2020-12-14 at 11 28 09

* Text in sidebar after schedule republishing.  
Screenshot 2020-12-14 at 11 28 39

_Classic editor_
* `Publish` to `Republish` in the Submitbox (title and button). 
Screenshot 2020-12-10 at 16 36 54

* `Schedule` to `Schedule republish` in the Submitbox.
Screenshot 2020-12-10 at 16 37 33

* Notice after schedule republishing.
Screenshot 2020-12-10 at 16 38 29

## Summary

This PR can be summarized in the following changelog entry:

* Changes copies in the post edit screen when a post is intended for republishing.

## Relevant technical choices:

* To change the copies in the classic editor, we hook into the `gettext` and `gettext_with_context` filters.
* To change the copies in the block editor, we use `setLocaleData`.
* We enqueue js script in the classic editor, because this is used when copies of the buttons are changed for example when you change between "regular" publishing and schedule publishing. 
* The ["Use new functions" commit](https://github.com/Yoast/duplicate-post/pull/65/commits/a9c39b0b06144a312c01da3d0f25cd0b0ceb08ce#diff-2fa4a5247864116d53349920d1d09a6de17a9736cdbce30e31cadf92946a0638L223-R223) is reverted in the last commit, because it caused errors in `admin-ajax.php`
Screenshot 2020-12-14 at 11 02 42, so after some discussion I decided to use the old method, which works and does not give any errors. When the code as a whole is reviewed at the end of this `Rewrite & Republish` project, it might be a good idea to revise this decision. 
* the minimum supported WP version is now 5.5, Duplicate Post from now on will support the latest two WP versions. [added by @enricobattocchi]
* consistently with the minimum WP supported version, the mimimum PHP supported version is now 5.6.20 [added by @enricobattocchi]

### Test instructions for the acceptance test before the PR gets merged
This PR can be acceptance tested by following these steps:

**Note: The following steps should be done for posts, pages and custom post types (e.g. Book or Movie from the Yoast Test helper).**

_Block editor_
* Make sure the Classic Editor is disabled. 
* Pick a published post and create a `Rewrite & Republish` copy using a link (not the R&R bulk action).
* In the following edit screen, see that (1) the publish button on the top right of the page states  `Republish` and (2) the test below "Visibility" in the Gutenberg sidebar states `Republish` [design](https://www.sketch.com/s/2ec4254a-9529-4616-abe7-f6262c2ed6dd/a/G1OaPb/play) (also see screenshots above). 
* Click the `Republish` button in the top right and see that the publish button states `Republish`, and the pre-publish copy matches the design, including the `Do you want to double-check your changes before merging?` (which does not do anything yet) [design](https://www.sketch.com/s/2ec4254a-9529-4616-abe7-f6262c2ed6dd/a/O7bp3k/play). 
* Click the `Cancel` button and change the publishing time to a time in the future.
* See the publish button on the top right corner now states `Schedule Republish`.
* Click the `Schedule Republish` in the top right and see that the button still states `Schedule Republish`, and the pre-publish copy matches the design, including the `Do you want to double-check your changes before merging?` (which does not do anything yet) [design](https://www.sketch.com/s/2ec4254a-9529-4616-abe7-f6262c2ed6dd/a/E115yd/play). 
* Click the `Schedule Republish` of the pre-publish check and see the copy matched the design `<title>, the rewritten post, is now scheduled to replace the original post. It will be published on <date> <time>.`. Check that the link leads to the frontend page of the post [design](https://www.sketch.com/s/2ec4254a-9529-4616-abe7-f6262c2ed6dd/a/qqq1Ed/play).
* Edit a post that is not intended for republishing and perform the steps mentioned above (for the block editor). See none of the copies are changed to `Republish` or `Schedule Republish`.

_Classic editor_
* Activate the Classic Editor. 
* Pick a published post and create a `Rewrite & Republish` copy using a link (not the R&R bulk action).
* In the following edit screen, see that the title and button in the Submitbox state`Republish` [design](https://www.sketch.com/s/2ec4254a-9529-4616-abe7-f6262c2ed6dd/a/rKGAbd/play). 
* Edit the publishing time to a time in the future and click `Save Draft`.
* See that the title is still `Republish` and the button now states `Schedule Republish` [design](https://www.sketch.com/s/2ec4254a-9529-4616-abe7-f6262c2ed6dd/a/9vYOr0/play). 
* Click the `Schedule Republish` button and see a dismissible notice (green) stating `This rewritten post <title> is now scheduled to replace the original post. It will be published on <date> <time>.` [design](https://www.sketch.com/s/2ec4254a-9529-4616-abe7-f6262c2ed6dd/a/Pyl8YL/play).
* Edit a post that is not intended for republishing and perform the steps mentioned above (for the classic editor). See none of the copies are changed to `Republish` or `Schedule Republish`.

_Test the translation_
To test the WP admin translation bug (see https://github.com/Yoast/duplicate-post/pull/65#issuecomment-746010303 ) is solved:
- in the WP admin > Settings > General, install and set a language other than English (if you don’t have one already)
- in the WP admin > Users > Profile, set the new language for your user profile
- observe the WP admin correctly switches to the new language 
- observe the overridden translations still work e.g. `Publish` => `Republish` and `Schedule` => `Schedule republish`

### Test instructions for QA when the code is in the RC

* [ ] QA should use the same steps as above.

Not applicable: a PR with extensive test instructions for the whole Rewrite and Republish feature (or for the whole plugin, to check for regressions etc.) will follow.

QA can test this PR by following these steps:

*

## Impact check

This PR affects the following parts of the plugin, which may require extra testing:

*

## UI changes

* [ ] This PR changes the UI in the plugin. I have added the 'UI change' label to this PR.

## Documentation

* [ ] I have written documentation for this change.

## Quality assurance

* [x] I have tested this code to the best of my abilities
* [x] I have added unittests to verify the code works as intended

Fixes #

### CODE REVIEW

**Review by enricobattocchi — COMMENTED** [review]
CR :construction: 
Use of a very recent WP function for a non-key functionality... maybe we can find a workaround?

File: config/webpack/paths.js [hunk 1]
```diff
@@ -7,6 +7,7 @@ const jsSrcPath = path.resolve( "js", "src" );
 // Entries for webpack to make bundles from.
 const entry = {
 	"duplicate-post-edit": "./duplicate-post-edit-script.js",
+	"duplicate-post-strings": "./ui/duplicate-post-strings.js",
```
[the remark this answers is not in the document]
Comment (karlijnbok) [reply config/webpack/paths.js:10]: Don't you think we will introduce more files in the future? We could of course also introduce the folder then. :)

File: js/src/ui/duplicate-post-strings.js [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+import { __, setLocaleData } from "@wordpress/i18n";
+
+const republishStrings = {
+    'Publish'  : __( 'Republish', 'duplicate-post' ),
+    'Publish…' : __( 'Republish', 'duplicate-post' ),
+
+    'Are you ready to publish?':
+        __( 'Are you ready to republish your post?', 'duplicate-post' ),
+    'Double-check your settings before publishing.':
+        createInterpolateElement(
```
[the remark this answers is not in the document]
Comment (karlijnbok) [reply js/src/ui/duplicate-post-strings.js:11]: For context: the reason the `createInterpolateElement` is used, is because some html needed to be added to the copy, i.e., line breaks and a link.

File: config/webpack/paths.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (enricobattocchi) [reply config/webpack/paths.js:10]: I'm not sure we'll have to, this is a small plugin :-D

[the remark this answers is not in the document]
Comment (karlijnbok) [reply config/webpack/paths.js:10]: Alright, I will remove it!

File: js/src/ui/duplicate-post-strings.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (enricobattocchi) [reply js/src/ui/duplicate-post-strings.js]: OK!
We decided to support only the latest two WP versions, so this will stay like this and the minimum WP version will be bumped.

**Comment by enricobattocchi** [comment]
CR OK!
Adding a commit to bump the minimum PHP supported version to the same one of WP 5.5 (which is the new minimum WP supported version), and to remove the check with PHP 5.5 from the Travis matrix (speeding up the build)

File: src/ui/class-post-submitbox.php [hunk 3]
```diff
@@ -123,4 +153,109 @@ public function add_rewrite_and_republish_post_button( $post = null ) {
...
+	 *
+	 * @return string The to-be-used copy of the text.
+	 */
+	public function change_schedule_strings_classic_editor( $translation, $text, $context ) {
+		if ( ! $this->should_change_rewrite_republish_copy( \get_post() ) ) {
+			return $text;
+		}
+
+		if ( $text === 'Schedule' ) {
+			return \__( 'Schedule republish', 'duplicate-post' );
```
Comment (afercia) [inline src/ui/class-post-submitbox.php:192]: The other two occurrences of `Schedule Republish` have an uppercase `R`. For consistency, they should all use the same casing. Note: While WordPress traditionally uses title case, at Yoast we generally prefer sentence case. Personally, I have no strong preferences in this case but it would be good to make them consistent. See also the related test.

File: tests/ui/class-post-submitbox-test.php [hunk 4]
```diff
@@ -358,4 +400,432 @@ public function test_add_rewrite_and_republish_post_button_not_publish() {
...
+		Monkey\Functions\expect( '\get_post' )
+			->once()
+			->andReturn( $post );
+
+		$this->instance->expects( 'should_change_rewrite_republish_copy' )
+			->with( $post )
+			->once()
+			->andReturnTrue();
+
+		$this->assertEquals( $this->instance->change_schedule_strings_classic_editor( '', $text, '' ), 'Schedule republish' );
```
Comment (afercia) [inline tests/ui/class-post-submitbox-test.php:495]: See comment about the uppercase/lowercase `R`.

File: src/ui/class-post-submitbox.php [hunk 5]
```diff
@@ -123,4 +153,109 @@ public function add_rewrite_and_republish_post_button( $post = null ) {
...
+	 * @return array[] The to-be-used messages.
+	 */
+	public function change_scheduled_notice_classic_editor( $messages ) {
+		$post = \get_post();
+		if ( ! $this->should_change_rewrite_republish_copy( $post ) ) {
+			return $messages;
+		}
+
+		$permalink      = \get_permalink( $post->ID );
+		$scheduled_date = \get_the_time( \get_option( 'date_format' ), $post );
```
Comment (afercia) [inline src/ui/class-post-submitbox.php:212]: According to [the design](https://www.sketch.com/s/2ec4254a-9529-4616-abe7-f6262c2ed6dd/a/Pyl8YL/play), the publish date should display also the publish _time_. See also the related test.

File: src/ui/class-post-submitbox.php [hunk 6]
```diff
@@ -123,4 +153,109 @@ public function add_rewrite_and_republish_post_button( $post = null ) {
...
+
+		if ( $post->post_type === 'post' ) {
+			$messages['post'][9] = \sprintf(
+			/* translators: 1: The post title with a link to the frontend page, 2: The scheduled date. */
+				\esc_html__(
+					'This rewritten post %1$s is now scheduled to replace the original post. It will be published on %2$s',
```
Comment (afercia) [inline src/ui/class-post-submitbox.php:218]: According to the design, the notice text should end with a period.

File: src/ui/class-post-submitbox.php [hunk 7]
```diff
@@ -123,4 +153,109 @@ public function add_rewrite_and_republish_post_button( $post = null ) {
...
+				'<strong>' . $scheduled_date . '</strong>'
+			);
+			return $messages;
+		}
+
+		if ( $post->post_type === 'page' ) {
+			$messages['page'][9] = \sprintf(
+					/* translators: 1: The page title with a link to the frontend page, 2: The scheduled date. */
+				\esc_html__(
+					'This rewritten page %1$s is now scheduled to replace the original page. It will be published on %2$s',
```
Comment (afercia) [inline src/ui/class-post-submitbox.php:231]: According to the design, the notice text should end with a period.

**Comment by afercia** [comment]
Acceptance 🚧 

I'm getting lots of PHP notices:

```
Notice: Trying to get property 'base' of non-object in /var/www/html/wp-content/plugins/duplicate-post/src/class-permissions-helper.php
```

at line 112 and 133. If you have the Query Monitor plugin activated, you should see it is highlighted in the admin bar and it reports 240 notices

Screenshot 2020-12-15 at 11 15 13

It it is not activated, you should get the notices printed out directly onto the page:

Screenshot 2020-12-15 at 11 15 36

If you don't get any notice, please do check the PHP error report level set in your local environment.

Regarding the reason of the notices, turns out that these two callback methors call `should_change_rewrite_republish_copy()`:

```
\add_filter( 'gettext', [ $this, 'change_republish_strings_classic_editor' ], 10, 2 );
\add_filter( 'gettext_with_context', [ $this, 'change_schedule_strings_classic_editor' ], 10, 3 );
```

`should_change_rewrite_republish_copy()` uses internally other two methods: `$permissions_helper->is_edit_post_screen()` and `$permissions_helper->is_new_post_screen()`. They both use `get_current_screen()` and check for the screen object properties without checking if it's null first.

However, the `gettext` and `gettext_with_context` filters run very early and at that point, the current screen object isn't set yet. It is null. The current screen object is set only on the `current_screen` hook and following hooks.

Two more minor things, see previous comments:
- uppercase / lowercase `R` inconsistency in the string 'Schedule Republish'
- the notice text should display also the publish time

File: js/src/duplicate-post-strings.js [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+};
+
+for ( const original in republishStrings ) {
+    setLocaleData( {
+        [ original ]: [
+            republishStrings[ original ],
+            'duplicate-post'
+        ]
+    } );
+}
```
Comment (afercia) [inline js/src/duplicate-post-strings.js:44]: Indentation across the whole file is made with spaces. It should be made with tab characters.

**Comment by afercia** [comment]
Spotted one more thing: Indentation across the whole `js/src/duplicate-post-strings.js` file is made with spaces. It should be made with tab characters.

**Comment by afercia** [comment]
Spotted (a bit late) a major issue: when switching the WP admin to a language other than English, the admin stays in English:
- in the WP admin > Settings > General, install and set a language other than English (if you don’t have one already)
- in the WP admin > Users > Profile, set the new language as your user profile
- observe the WP admin stays in English 😱 
- deactivate Duplicate post, observe the WP admin switches to the new language

It appears this happens because the methods hooked to the `gettext` and `gettext_with_context` _filters_ used to change the buttons text don’t return the first passed argument, which is the translated text. Filters should always return the original passed argument.

**Comment by enricobattocchi** [comment]
CR on the additions OK! :+1:

**Comment by enricobattocchi** [comment]
Acceptance test OK! :+1:
